File: dashboard/src/lib/onboarding.ts
import axios from "axios";
import type { ApiEnvelope } from "./api";
import { gatewayURL } from "./runtime-config";

function onboardingHeaders(adminSecret?: string): Record<string, string> {
  const headers: Record<string, string> = {};
  const supabaseUserId = localStorage.getItem("MOS_SUPABASE_USER_ID");
  if (supabaseUserId) {
    headers["X-Supabase-User-Id"] = supabaseUserId;
  }
  if (adminSecret) {
    headers["X-Admin-Secret"] = adminSecret;
  }
  return headers;
}

const client = axios.create({
  baseURL: `${gatewayURL()}/api/v1`,
});

export type ApiKeySummary = {
  id: number;
  name: string;
  key_prefix: string;
  tenant_id: number;
  org_id: number;
  user_id: number;
  revoked_at?: string | null;
};

export type OnboardingProfile = {
  tenant_id: number;
  org_id: number;
  user_id: number;
  email: string;
  display_name?: string | null;
  tenant_code?: string | null;
  org_code?: string | null;
  api_keys: ApiKeySummary[];
};

export type ApiKeyIssue = {
  tenant_id: number;
  org_id: number;
  user_id: number;
  api_key_id: number;
  api_key: string;
  key_prefix: string;
};

export type TenantBootstrapResult = ApiKeyIssue;

export async function fetchOnboardingProfile(): Promise<OnboardingProfile> {
  const { data } = await client.get<ApiEnvelope<OnboardingProfile>>("/onboarding/me", {
    headers: onboardingHeaders(),
  });
  return data.data;
}

export async function applyApiKey(payload: {
  email: string;
  display_name?: string;
  name?: string;
}): Promise<ApiKeyIssue> {
  const { data } = await client.post<ApiEnvelope<ApiKeyIssue>>(
    "/onboarding/api-key/apply",
    {
      email: payload.email,
      display_name: payload.display_name,
      name: payload.name ?? "dashboard",
    },
    { headers: onboardingHeaders() },
  );
  return data.data;
}

export async function createTenant(
  body: {
    tenant_code: string;
    tenant_name: string;
    org_code: string;
    org_name: string;
    email: string;
    display_name?: string;
    api_key_name?: string;
    supabase_user_id?: string;
  },
  adminSecret: string,
): Promise<TenantBootstrapResult> {
  const { data } = await client.post<ApiEnvelope<TenantBootstrapResult>>(
    "/onboarding/tenant",
    body,
    { headers: onboardingHeaders(adminSecret) },
  );
  return data.data;
}

export async function listApiKeys(): Promise<ApiKeySummary[]> {
  const apiKey = localStorage.getItem("MOS_API_KEY") || "";
  const tenantId = localStorage.getItem("MOS_TENANT_ID") || "1";
  const orgId = localStorage.getItem("MOS_ORG_ID") || "1";
  const { data } = await client.get<ApiEnvelope<ApiKeySummary[]>>("/onboarding/api-keys", {
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "X-Tenant-Id": tenantId,
      "X-Org-Id": orgId,
    },
  });
  return data.data;
}

export async function createApiKey(name: string): Promise<ApiKeyIssue> {
  const apiKey = localStorage.getItem("MOS_API_KEY") || "";
  const tenantId = localStorage.getItem("MOS_TENANT_ID") || "1";
  const orgId = localStorage.getItem("MOS_ORG_ID") || "1";
  const { data } = await client.post<ApiEnvelope<ApiKeyIssue>>(
    "/onboarding/api-keys",
    { name },
    {
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "X-Tenant-Id": tenantId,
        "X-Org-Id": orgId,
      },
    },
  );
  return data.data;
}

File: backend/src/memory_engine/schemas/onboarding.py
"""Dashboard onboarding: API keys and tenant self-service."""

from datetime import datetime

from pydantic import BaseModel, Field


class ApiKeyApplyRequest(BaseModel):
    """Apply for a new API key (Supabase user or admin-assisted)."""

    name: str = Field(default="dashboard", max_length=128)
    email: str | None = Field(default=None, max_length=320)
    display_name: str | None = Field(default=None, max_length=255)


class ApiKeyCreateRequest(BaseModel):
    """Create an additional API key for the current tenant (authenticated)."""

    name: str = Field(default="default", max_length=128)


class ApiKeySummary(BaseModel):
    """API key metadata (never includes plaintext secret)."""

    id: int
    name: str
    key_prefix: str
    tenant_id: int
    org_id: int
    user_id: int
    revoked_at: datetime | None = None
    expires_at: datetime | None = None
    create_time: datetime | None = None


class ApiKeyIssueResponse(BaseModel):
    """Newly issued API key — plaintext shown once."""

    tenant_id: int
    org_id: int
    user_id: int
    api_key_id: int
    api_key: str
    key_prefix: str


class OnboardingProfileResponse(BaseModel):
    """Current Supabase-linked Memory Engine user profile."""

    tenant_id: int
    org_id: int
    user_id: int
    email: str
    display_name: str | None
    tenant_code: str | None = None
    org_code: str | None = None
    api_keys: list[ApiKeySummary] = Field(default_factory=list)
